### ISSUE TITLE
File Attachment Doesn't Work with PDF

### ISSUE THREAD

**Issue by rb81** [BoltAI/BoltAI#148]
**Describe the bug**
It appears that the file attachment feature doesn't work when attaching a PDF and adding the full document to the context window.

**To Reproduce**
Add Document (Full Content).

**Expected behavior**
The model is expected to receive the content of the file within the current context window.

**Version (please complete the following information):**
 - macOS: 14.6.1
 - app version: 1.18.3 (build 97)

**Comment by longseespace** [comment]
It really depends on the PDF I think. PDF is the hardest file format to work with. If the file is not sensive, can you share it with me so I could debug it better.

Thanks

**Comment by rb81** [comment]
@longseespace - It's a messy PDF honestly, but works with my custom RAG.

[Andersen-v-Stability-AI-Order-on-Motions-to-Dismiss-8-12-2024.pdf](https://github.com/user-attachments/files/16627553/Andersen-v-Stability-AI-Order-on-Motions-to-Dismiss-8-12-2024.pdf)

**Comment by longseespace** [comment]
I'm not sure what wrong but I couldn't reproduce the issue. But there must be an issue I think, since this is the second report I received regarding this issue. I'm adding the ability to view the content that BoltAI parsed.

Quick question: how did you attach these documents? Did you open from Finder, drag to Dock, or start new chat and select the file?

**Comment by rb81** [comment]
I added the document using the (+) in the chat window.

I tried this several times earlier today and it didn't work, but, oddly enough, it works now. I tested it in various scenarios, but couldn't reproduce the issue. Could it be that Bolt requires time to process the document before it is properly included in the context window?

**Comment by tonilampela** [comment]
With latest Bolt update (1.27.1) I noticed that any queries made to attached document seems to be heavily hallucinated. Previous version worked somewhat OK, except that I too sometimes saw responses that Bolt "lost" knowledge of any attachments. I'm using local ollama and llama3.2 as LLM.

**Comment by longseespace** [comment]
> I added the document using the (+) in the chat window.
> 
> I tried this several times earlier today and it didn't work, but, oddly enough, it works now. I tested it in various scenarios, but couldn't reproduce the issue. Could it be that Bolt requires time to process the document before it is properly included in the context window?

No. The PDF is processed right away. But maybe there was an issue with including it in the context. Do you still run into this issue? @rb81

**Comment by longseespace** [comment]
> With latest Bolt update (1.27.1) I noticed that any queries made to attached document seems to be heavily hallucinated. Previous version worked somewhat OK, except that I too sometimes saw responses that Bolt "lost" knowledge of any attachments. I'm using local ollama and llama3.2 as LLM.

Try toggle the option "Append document content to System Prompt" (screenshot below). Alternatively, try to change the "Context Limit" to all messages.

![CleanShot 2024-11-12 at 15 25 41@2x](https://github.com/user-attachments/assets/202eda4b-d22a-4529-9ca3-191751534375)

There are 2 approaches to attach document content to the chat context:

1. **To the System Prompt.** this was how BoltAI works pre v1.27.1. Basically BoltAI attaches all document content to the System Prompt. The good thing about this is you can cache all document content if available (Claude Prompt Caching). One issue is it might lose the immediate context. LLMs might not understand when you ask "what about this?" because your prompt is separated from the document content.

2. **To each individual message:** this is new in v1.27.1. It's pretty straightforward. However since the document content is attached to each individual message, it follows the Context Limit configuration (read more: https://boltai.com/docs/chat-ui/chat-configuration#context-limit). So for example, if you set the Context Limit to last 5 messages, and your document is attached to the message that is not the last 5 messages then it might not be included in the chat context.

I'm not sure how best to communicate this tradeoff  in the app in terms of UX. Each of these approaches has its own pros and cons.
